File: [1353] Maximum Number of Events That Can Be Attended.cpp
class Solution {
public:
    int maxEvents(vector<vector<int>>& events)
    {
        //greedy: sort by start time and end time.
        //attend the one started first.
        //evict all expired the next day and attend the one ends first
        sort(events.begin(), events.end());
        priority_queue<int, vector<int>, greater<int>>pq;
        int i = 0;
        int ret = 0;
        for (int day=1; day<=100000; day++)//each day see who got in scope
        {
            while (i<events.size() && events[i][0]<=day)//if you happen today, add to pq
            {
                pq.push(events[i][1]);
                i++;
            }
            while (!pq.empty() && pq.top() < day)//evict all those expired
                pq.pop();
            if (!pq.empty())
            {
                pq.pop();//take the one ends first and add count
                ret++;
            }
        }
        return ret;
    }
};

File: LC/[1353] Maximum Number of Events That Can Be Attended.py
from heapq import heappush, heappop
class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        # greedy, attend the one starts the first and add all those available to a PQ.
        # the next day. add newly available events into PQ, evict those expired.
        # pop the top one in PQ to attend and continue

        max_day=pow(10, 5)+1
        mnq=[]
        events.sort() # sort by st, ed in increasing order
        ii=0
        cnt=0
        for day in range(1, max_day):
            while ii < len(events) and events[ii][0]<= day:
                heappush(mnq, events[ii][1])
                ii+=1
            
            while len(mnq) > 0 and mnq[0] < day:
                heappop(mnq)

            if len(mnq) > 0:
                cnt += 1
                heappop(mnq)

        return cnt
